cmake/extensions.cmake

```diff
@@ -2261,3 +2263,369 @@ function(target_byproducts)
                      COMMENT "Logical command for additional byproducts on target: ${TB_TARGET}"
   )
 endfunction()
+
+########################################################
```

Comment:
can we split this out into its own file.

Reply:
we could, but are we sure that we want to do that ?

Today, whenever one sees a CMake function or macro in use inside the that is not vanilla CMake, then there is one place to look, and that is `cmake/extensions.cmake`.

That makes it easy to lookup extension functions.

Only exception are areas where functions are used internally, such as: `cmake/flash/CMakeLists.txt` but those functions are not use outside that file, afaik.

We could of course establish a new scheme, like:
- `cmake/extensions/dt-extensions.cmake`
- `cmake/extensions/library-extensions.cmake`
  - for library related extensions, like `zephyr_library()`,  `zephyr_library_sources`, `zephyr_sources`, ...
- `cmake/extensions/generic-extensions.cmake`
  - for generic CMake functions with a twist, like `target_sources_ifdef()`, `add_subdirectory_ifdef()`.
- `cmake/extensions/misc-extensions.cmake`
  - There always needs to be a misc.

but doing such split (which would be outside the scope of this PR) will sometimes make it harder to know where a given function belongs (most functions will be easy, but not all).

Having only devicetree functions in a separate file just makes it inconsistent where functions are placed, imho.

If it is because you believe this feature could be reusable outside Zephyr scope, like with linker generator, then I see the point and a use-case.

Reply:
the thought was that this could be usable outside of zephyr.

Reply:
> the thought was that this could be usable outside of zephyr.

makes sense, but those functions are useless without the generated `<build>/zephyr/dts.cmake` file which is created here https://github.com/zephyrproject-rtos/zephyr/blob/aecb7fc1b488ac3d017447d324b1e78e793f7620/scripts/dts/dts_cmake.py based on the `edt.pickle` file, https://github.com/zephyrproject-rtos/zephyr/blob/aecb7fc1b488ac3d017447d324b1e78e793f7620/scripts/dts/dts_cmake.py#L34

The `edt.pickle` file is created in the `gen_defines.py` script here: https://github.com/zephyrproject-rtos/zephyr/blob/33d5be865653dbc3628a86d1447fca8de6ce398e/scripts/dts/gen_defines.py#L929

so just placing those dt_* functions in its own CMake file doesn't really make them reusable.

If we really want to make it easier to reuse the devicetree --> CMake infrastructure outside Zephyr, then perhaps we should see if this could be streamlined with reusability in mind for all the parts involved:
- [\<zephyr\>/cmake/dts.cmake](https://github.com/zephyrproject-rtos/zephyr/blob/aecb7fc1b488ac3d017447d324b1e78e793f7620/cmake/dts.cmake)
- [\<zephyr\>/scripts/dts/dts_cmake.py](https://github.com/zephyrproject-rtos/zephyr/blob/aecb7fc1b488ac3d017447d324b1e78e793f7620/scripts/dts/dts_cmake.py)
- [\<zephyr\>/scripts/dts/gen_defines.py](https://github.com/zephyrproject-rtos/zephyr/blob/aecb7fc1b488ac3d017447d324b1e78e793f7620/scripts/dts/gen_defines.py)
- [\<zephyr\>/cmake/extensions.cmake](https://github.com/zephyrproject-rtos/zephyr/blob/aecb7fc1b488ac3d017447d324b1e78e793f7620/cmake/extensions.cmake), devicetree part only.

That sounds very reasonable to me, but I think it would be outside the scope of this PR.

Reply:
I agree with @tejlmand . If someone wants to use it outside of Zephyr they can pick up the patch at that point and make it do what they need. I'm not sure it's really necessary for this PR.